rust/network/src/receiver.rs

```diff
@@ -0,0 +1,88 @@
+use crate::error::NetworkError;
+use async_trait::async_trait;
+use bytes::Bytes;
+use futures::stream::SplitSink;
+use futures::stream::StreamExt as _;
+use log::{debug, info, warn};
+use std::error::Error;
+use std::net::SocketAddr;
+use tokio::net::{TcpListener, TcpStream};
+use tokio_util::codec::{Framed, LengthDelimitedCodec};
+
+#[cfg(test)]
+#[path = "tests/receiver_tests.rs"]
+pub mod receiver_tests;
+
+/// Convenient alias for the writer end of the TCP channel.
+pub type Writer = SplitSink<Framed<TcpStream, LengthDelimitedCodec>, Bytes>;
+
+#[async_trait]
+pub trait MessageHandler: Clone + Send + Sync + 'static {
+    /// Defines how to handle an incoming message. A typical usage is to define a `MessageHandler` with a
+    /// number of `Sender<T>` channels. Then implement `dispatch` to deserialize incoming messages and
+    /// forward them through the appropriate delivery channel. Then `writer` can be used to send back
+    /// responses or acknowledgements to the sender machine (see unit tests for examples).
```

Comment:
Writer can be dropped after the handler takes over. It should be for the handler to define that behavior